storage-provider/server/src/main.rs

```
@@ -390,63 +367,77 @@ impl Server {
+                .map(|result| ("HTTP Storage", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_pipeline(
+                Arc::new(pipeline_state),
+                pipeline_rx,
+                cancellation_token.child_token(),
+            )
+            .map(|result| ("Pipeline", result.map_err(ServerError::from))),
+        );
+        tasks.spawn(
+            start_indexer(indexer_state, indexer_rx, cancellation_token.child_token())
+                .map(|result| ("Indexer", result.map_err(ServerError::from))),
+        );
+        tracing::info!("Successfully launched all sub-services, ready for work!");
 
-        // Exit with error
-        upload_result??;
-        rpc_task??;
-        pipeline_task??;
-        p2p_task??;
-        indexer_task??;
-
-        Ok(())
+        // Keep the first error around as the "canonical return",
+        // since we can't return multiple values
+        let mut error = None;
+        loop {
+            tokio::select! {
+                result = tasks.join_next() => {
+                    match result {
+                        Some(Ok((task_name, task_result))) => {
+                            match task_result {
+                                Ok(()) => tracing::info!("{task_name} finished successfully!"),
+                                Err(err) => {
+                                    tracing::error!("{task_name} finished with error: {err}");
+                                    tracing::error!("Cancelling remaining tasks...");
+                                    if error.is_none() {
+                                        error = Some(err);
+                                    }
+                                    cancellation_token.cancel();
+                                },
+                            }
+                        }
+                        Some(Err(err)) => {
+                            tracing::error!("Failed to join task with error: {err}");
```

Why are we ignoring this error with logging and not saving it to `error = `?

---

What do you mean by, ignoring the error?

---

Good catch. This should also set the error if it is None